### PR TITLE
feat: add SearXNG meta search engine tool integration

### PR DESCRIPTION
**Pull request by stoykovstoyk**
#### Problem
PentAGI lacked integration with Searxng, a popular privacy-focused meta search engine that aggregates results from multiple search engines. Users requested the ability to add Searxng as an additional search option alongside existing providers like Google, DuckDuckGo, Tavily, and Perplexity.

#### Solution
Implemented comprehensive Searxng search tool integration with the following key components:

 - Created backend/pkg/tools/searxng.go with full tool implementation supporting configurable search parameters
 - Added Searxng environment variables to the configuration system for flexible deployment
 - Implemented proper HTTP client with timeout and proxy support for reliable API communication
- Added comprehensive error handling and logging for debugging and monitoring
 - Created database integration for search result tracking and analytics
 - Developed extensive unit tests with mock servers to ensure reliability
 - Updated documentation to reflect the new search capability

Closes #

### Type of Change

- [ ] 🐛 Bug fix (non-breaking change which fixes an issue)
- [x] 🚀 New feature (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] 📚 Documentation update
- [ ] 🔧 Configuration change
- [x] 🧪 Test update
- [ ] 🛡️ Security update

### Areas Affected

- [ ] Core Services (Frontend UI/Backend API)
- [ ] AI Agents (Researcher/Developer/Executor)
- [ ] Security Tools Integration
- [ ] Memory System (Vector Store/Knowledge Base)
- [ ] Monitoring Stack (Grafana/OpenTelemetry)
- [ ] Analytics Platform (Langfuse)
- [x] External Integrations (LLM/Search APIs)
- [x] Documentation
- [ ] Infrastructure/DevOps

#### Test Configuration
```yaml
PentAGI Version: Latest master branch
Docker Version: 24.0.0+
Host OS: Linux
LLM Provider: Not required for this feature
Enabled Features: [ftester for testing, mock mode]
```

#### Test Steps
1. Build the backend: cd backend && go build ./...
2. Run unit tests: cd backend && go test -v ./pkg/tools/
3. Verify all tests pass including edge cases and error scenarios

#### Test Results
All unit tests pass successfully:

- [x] TestNewSearxngTool: Tool creation and initialization
- [x] TestSearxngToolIsAvailable: Availability checking logic
- [x] TestSearxngToolHandleWithInvalidArgs: Invalid argument handling
- [x] TestSearxngToolHandleWithValidArgs: Valid search execution with mock server
- [x] TestSearxngToolHandleWithServerError: Server error handling
- [x] TestSearxngToolHandleWithInvalidURL: Invalid URL handling
- [x] TestSearxngToolHandleWithNoResults: No results scenarios
- [x] Build verification shows no compilation errors and all existing functionality remains intact.

### Security Considerations
The Searxng integration enhances security by providing users with a privacy-focused search option that can be self-hosted. Key security considerations:

 - Network Security: The tool supports proxy configuration for network isolation
 - Input Validation: All search queries are validated before API calls
 - Error Handling: Comprehensive error handling prevents information leakage
 - No New Dependencies: Uses existing HTTP client and database libraries
 - Self-Hostable Option: Users can run their own Searxng instance for maximum privacy
 - Configuration Security: Environment variables follow existing security patterns

### Performance Impact
The Searxng tool is designed with performance in mind:

 - Configurable Timeouts: Default 30-second timeout prevents hanging requests
 - Connection Reuse: Uses standard HTTP client for efficient connection management
 - Proxy Support: Optional proxy configuration for enterprise environments
 - Result Caching: Database integration allows for result tracking and potential caching
 - Lightweight Implementation: Minimal overhead, follows existing search tool patterns
 - Async Processing: Non-blocking design that doesn't impact other system components
 - Performance testing with mock servers confirmed sub-second response times for typical search queries.

### Documentation Updates

- [x] README.md updates
- [ ] API documentation updates
- [x] Configuration documentation updates
- [ ] GraphQL schema updates
- [x] Other:   EULA.md updates

### Deployment Notes
New Environment Variables
```
# Required: Searxng instance URL
SEARXNG_URL=http://localhost:8080

# Optional: Search configuration
SEARXNG_CATEGORIES=general          # Default: "general"
SEARXNG_LANGUAGE=en                # Default: "en" 
SEARXNG_SAFESearch=0               # Default: "0"
SEARXNG_TIME_RANGE=                # Optional: time range filter
SEARXNG_PROXY_URL=                 # Optional: proxy URL
```

#### Code Quality
- [x] My code follows the project's coding standards
- [x] I have added/updated necessary documentation
- [x] I have added tests to cover my changes
- [x] All new and existing tests pass
- [x] I have run `go fmt` and `go vet` (for Go code)
- [ ] I have run `npm run lint` (for TypeScript/JavaScript code)

#### Security
- [x] I have considered security implications
- [x] Changes maintain or improve the security model
- [x] Sensitive information has been properly handled

#### Compatibility
- [x] Changes are backward compatible
- [ ] Breaking changes are clearly marked and documented
- [x] Dependencies are properly updated

#### Documentation
- [x] Documentation is clear and complete
- [x] Comments are added for non-obvious code
- [x] API changes are documented

### Additional Notes
The Searxng tool implementation follows the exact same patterns as existing search tools (Google, DuckDuckGo, Tavily, etc.) ensuring consistency in the codebase. The feature is optional and only activates when SEARXNG_URL is configured, making it safe for production deployment.

Key implementation highlights:

 - Uses existing SearchLogProvider interface for database integration
 - Follows established error handling and logging patterns
 - Includes comprehensive mock support for testing
 - Supports all Searxng API parameters with sensible defaults
 - Maintains compatibility with existing tool registry system

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -93,6 +93,14 @@ PERPLEXITY_API_KEY=
 PERPLEXITY_MODEL=
 PERPLEXITY_CONTEXT_SIZE=
 
+## SEARXNG search engine API
+SEARXNG_URL=https://searxng.org
```
Comment (asdek) [inline .env.example]: I think, by default it should be disabled because this value is using in `IsAvailable()` call

File: docker-compose.yml [hunk 2]
```diff
@@ -83,6 +83,10 @@ services:
       - OAUTH_GITHUB_CLIENT_SECRET=${OAUTH_GITHUB_CLIENT_SECRET:-}
       - DATABASE_URL=postgres://${PENTAGI_POSTGRES_USER:-postgres}:${PENTAGI_POSTGRES_PASSWORD:-postgres}@pgvector:5432/${PENTAGI_POSTGRES_DB:-pentagidb}?sslmode=disable
       - DUCKDUCKGO_ENABLED=${DUCKDUCKGO_ENABLED:-true}
+      - SEARXNG_URL=${SEARXNG_URL:-}
+      - SEARXNG_CATEGORIES=${SEARXNG_CATEGORIES:-}
+      - SEARXNG_LANGUAGE=${SEARXNG_LANGUAGE:-}
+      - SEARXNG_SAFESearch=${SEARXNG_SAFESearch:-}
```
Comment (asdek) [inline docker-compose.yml:89]: here are not all of variables passing to the container, SEARXNG_TIME_RANGE and SEARXNG_PROXY_URL should be pass too

File: backend/pkg/tools/searxng.go [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+		return "", fmt.Errorf("error unmarshaling search arguments: %w", err)
+	}
+
+	// Validate required parameters
+	if searchArgs.Query == "" {
+		return "", fmt.Errorf("query parameter is required")
+	}
+
+	// Log the search
+	searchLogID, err := s.searchLog.PutLog(ctx, database.MsgchainTypeAssistant, database.MsgchainTypeAssistant, database.SearchengineTypeSearxng, searchArgs.Query, "", s.taskID, s.subtaskID)
```
Comment (asdek) [inline backend/pkg/tools/searxng.go]: could you please [look it up](https://github.com/vxcontrol/pentagi/blob/master/backend/pkg/tools/google.go#L93-L104) this code for getting agent type information from context and after that will change this call?
also, will be better to save variables name convension similar like with other files in the package.

**Comment by asdek** [comment]
hello @stoykovstoyk 

LGTM, but I found some places for improving, would you make it and I'll merge it.

Thank you for contribution! 😉

**Comment by stoykovstoyk** [comment]
No problem, I will do it :)
What do you want me to improve?

File: docker-compose.yml (hunk 2 above)
Comment (stoykovstoyk) [reply docker-compose.yml:89]: I added the missing SEARXNG environment variables to docker-compose.yml

File: .env.example (hunk 1 above)
Comment (stoykovstoyk) [reply .env.example]: i disabled the SEARXNG_URL by default in .env.example

File: backend/pkg/tools/searxng.go (hunk 3 above)
Comment (stoykovstoyk) [reply backend/pkg/tools/searxng.go]: I updated logging to use agent context and follow naming conventions

**Comment by asdek** [comment]
@stoykovstoyk 

perfectly! your PR accepted, tnx again!
